java/core/src/main/java/co/worklytics/psoxy/storage/impl/ColumnarBulkDataSanitizerImpl.java

```diff
@@ -141,39 +167,53 @@ public byte[] sanitize(@NonNull InputStreamReader reader,
-                                    throw new Error("Cannot encode legacy PseudonymizedIdentity with reversibles as URL_SAFE_TOKEN");
-                                }
-                                return urlSafeTokenPseudonymEncoder.encode(identity.fromLegacy());
-                            } else {
-                                return urlSafeTokenPseudonymEncoder.encode(identity.asPseudonym());
-                            }
-                        } else {
-                            //JSON
-                            return objectMapper.writeValueAsString(identity);
-                        }
-                    } catch (JsonProcessingException e) {
-                        throw new RuntimeException(e);
-                    }
+            if (columnsToPseudonymize.contains(outputColumnName.toLowerCase())) {
+                return pseudonymizationFunction.apply(value, outputColumnName, pseudonymizer);
+            } else if (columnsToPseudonymizeWithScope.containsKey(outputColumnName.toLowerCase())) {
+                Pseudonymizer scopedPseudonymizer = pseudonymizer;
+                if (pseudonymizer.getOptions().getPseudonymImplementation() == PseudonymImplementation.LEGACY) {
+                    scopedPseudonymizer = pseudonymizers.computeIfAbsent(columnsToPseudonymizeWithScope.get(outputColumnName.toLowerCase()),
+                        scope -> pseudonymizerImplFactory.create(pseudonymizer.getOptions().withDefaultScopeId(scope)));
                 }
+
+                return pseudonymizationFunction.apply(value, outputColumnName, scopedPseudonymizer);
             }
             return value;
         };
 
+        BiFunction<String, String, String> applyTransformIfAny = (outputColumnName, value) -> {
+          if (columnsToTransform.containsKey(outputColumnName.toLowerCase())) {
+              ColumnarRules.FieldValueTransform transform = columnsToTransform.get(outputColumnName.toLowerCase());
+
+              if (transform.getFilterRegex() != null) {
+                  Pattern pattern = Pattern.compile(transform.getFilterRegex());
+                  Matcher matcher = pattern.matcher(value);
+                  if (matcher.matches()) {
+                      if (matcher.groupCount() > 0) {
+                          value = matcher.group(1);
+                      }
+                      if (transform.getOutputTemplate() != null) {
+                          if (transform.getOutputTemplate().contains("%s")) {
+                              value = String.format(transform.getOutputTemplate(), value);
+                          } else {
+                              throw new IllegalArgumentException("Invalid outputTemplate:" + transform.getOutputTemplate() + " for column " + outputColumnName + ". Must contain %s");
+                          }
+                      }
+                  } else {
+                      value = null;
+                  }
```

Comment:
Quick sample of what I mentioned, works and more versatile as can leverage full power of regexps, not just one single replacement
```
Matcher matcher = Pattern.compile("(?<username>.*)@worklytics.co").matcher("jose@worklytics.co");

if (matcher.matches() && transform.getOutputTemplate() != null) {
			value = matcher.replaceAll("${username}_github");
} else {
value = null;
}
```